In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt

In [2]:
applic = pd.read_csv("./data/data/applicant.csv")
loan=pd.read_csv("./data/data/loan.csv")

In [3]:
merged_data = loan.merge(applic,on=["applicant_id"])

In [4]:
column_names = ['loan_application_id', 'applicant_id', 'Months_loan_taken_for','Purpose', 'Principal_loan_amount','EMI_rate_in_percentage_of_disposable_income', 'Property','Has_coapplicant', 'Has_guarantor', 'Other_EMI_plans','Number_of_existing_loans_at_this_bank', 'Loan_history','Primary_applicant_age_in_years', 'Gender','Marital_status', 'Number_of_dependents', 'Housing','Years_at_current_residence', 'Employment_status','Has_been_employed_for_at_least', 'Has_been_employed_for_at_most','Telephone', 'Foreign_worker','Savings_account_balance','Balance_in_existing_bank_account_(lower_limit_of_bucket)','Balance_in_existing_bank_account_(upper_limit_of_bucket)','high_risk_applicant']

In [5]:
df = merged_data.reindex(columns=column_names)

In [6]:
df.head()

loan_application_id  applicant_id  Months_loan_taken_for  \
0  d68d975e-edad-11ea-8761-1d6f9c1ff461       1469590                      6   
1  d68d989e-edad-11ea-b1d5-2bcf65006448       1203873                     48   
2  d68d995c-edad-11ea-814a-1b6716782575       1432761                     12   
3  d68d99fc-edad-11ea-8841-17e8848060ae       1207582                     42   
4  d68d9a92-edad-11ea-9f3d-1f8682db006a       1674436                     24   

                Purpose  Principal_loan_amount  \
0  electronic equipment                1169000   
1  electronic equipment                5951000   
2             education                2096000   
3                  FF&E                7882000   
4           new vehicle                4870000   

   EMI_rate_in_percentage_of_disposable_income  \
0                                            4   
1                                            2   
2                                            2   
3                                            2   
4                                            3   

                                            Property  Has_coapplicant  \
0                                        real estate                0   
1                                        real estate                0   
2                                        real estate                0   
3  building society savings agreement/life insurance                0   
4                                                NaN                0   

   Has_guarantor Other_EMI_plans  ...  Years_at_current_residence  \
0              0             NaN  ...                           4   
1              0             NaN  ...                           2   
2              0             NaN  ...                           3   
3              1             NaN  ...                           4   
4              0             NaN  ...                           4   

             Employment_status  Has_been_employed_for_at_least  \
0  skilled employee / official                         7 years   
1  skilled employee / official                          1 year   
2         unskilled - resident                         4 years   
3  skilled employee / official                         4 years   
4  skilled employee / official                          1 year   

  Has_been_employed_for_at_most                              Telephone  \
0                           NaN  Registered under the applicant's name   
1                       4 years                                    NaN   
2                       7 years                                    NaN   
3                       7 years                                    NaN   
4                       4 years                                    NaN   

   Foreign_worker Savings_account_balance  \
0               1                     NaN   
1               1                     Low   
2               1                     Low   
3               1                     Low   
4               1                     Low   

   Balance_in_existing_bank_account_(lower_limit_of_bucket)  \
0                                                NaN          
1                                                  0          
2                                                NaN          
3                                                NaN          
4                                                NaN          

  Balance_in_existing_bank_account_(upper_limit_of_bucket) high_risk_applicant  
0                                                  0                         0  
1                                              2 lac                         1  
2                                                NaN                         0  
3                                                  0                         0  
4                                                  0                         1  

[5 rows x 27 columns]

In [7]:
numerical=['applicant_id', 'Months_loan_taken_for', 'Principal_loan_amount', 'EMI_rate_in_percentage_of_disposable_income', 'Has_coapplicant', 'Has_guarantor', 'Number_of_existing_loans_at_this_bank', 'Primary_applicant_age_in_years', 'Number_of_dependents', 'Years_at_current_residence', 'Foreign_worker', 'high_risk_applicant']

In [8]:
for i in numerical:
    print((df[i]<0).any())

False
False
False
False
False
False
False
False
False
False
False
False


In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 0 to 999
Data columns (total 27 columns):
 #   Column                                                    Non-Null Count  Dtype 
---  ------                                                    --------------  ----- 
 0   loan_application_id                                       1000 non-null   object
 1   applicant_id                                              1000 non-null   int64 
 2   Months_loan_taken_for                                     1000 non-null   int64 
 3   Purpose                                                   988 non-null    object
 4   Principal_loan_amount                                     1000 non-null   int64 
 5   EMI_rate_in_percentage_of_disposable_income               1000 non-null   int64 
 6   Property                                                  846 non-null    object
 7   Has_coapplicant                                           1000 non-null   int64 
 8   Has_guarantor                

In [10]:
df.isnull().sum()

loan_application_id                                           0
applicant_id                                                  0
Months_loan_taken_for                                         0
Purpose                                                      12
Principal_loan_amount                                         0
EMI_rate_in_percentage_of_disposable_income                   0
Property                                                    154
Has_coapplicant                                               0
Has_guarantor                                                 0
Other_EMI_plans                                             814
Number_of_existing_loans_at_this_bank                         0
Loan_history                                                  0
Primary_applicant_age_in_years                                0
Gender                                                        0
Marital_status                                                0
Number_of_dependents                    

In [11]:
df.nunique()

loan_application_id                                         1000
applicant_id                                                1000
Months_loan_taken_for                                         33
Purpose                                                        9
Principal_loan_amount                                        921
EMI_rate_in_percentage_of_disposable_income                    4
Property                                                       3
Has_coapplicant                                                2
Has_guarantor                                                  2
Other_EMI_plans                                                2
Number_of_existing_loans_at_this_bank                          4
Loan_history                                                   5
Primary_applicant_age_in_years                                53
Gender                                                         2
Marital_status                                                 4
Number_of_dependents     

<b>As we can see there is high variance in 'Principal_loan_amount','loan_application_id','applicant_id'

## Handling the missing values

In [12]:
df['Telephone'].value_counts()

Registered under the applicant's name    404
Name: Telephone, dtype: int64

In [13]:
df0 = df.copy()
df0.loc[df0['Telephone']=="Registered under the applicant's name", 'Telephone']=1

In [14]:
df0['Telephone']=df0['Telephone'].fillna(0)

In [15]:
df0['Telephone'].value_counts()

0    596
1    404
Name: Telephone, dtype: int64

In [16]:
for col in df0:
    # Check if the column is of object type
    if df0[col].dtype == 'object':
        # Impute with the most frequent value
        df0 = df0.fillna(df0[col].value_counts().index[0])

# Count the number of NaNs in the dataset and print the counts to verify
print(df0.isna().sum())

loan_application_id                                         0
applicant_id                                                0
Months_loan_taken_for                                       0
Purpose                                                     0
Principal_loan_amount                                       0
EMI_rate_in_percentage_of_disposable_income                 0
Property                                                    0
Has_coapplicant                                             0
Has_guarantor                                               0
Other_EMI_plans                                             0
Number_of_existing_loans_at_this_bank                       0
Loan_history                                                0
Primary_applicant_age_in_years                              0
Gender                                                      0
Marital_status                                              0
Number_of_dependents                                        0
Housing 

## Preprocessing the data

### categorical features are converted into numerical feature

In [17]:
# Import LabelEncoder
from sklearn.preprocessing import LabelEncoder

# Instantiate LabelEncoder
le = LabelEncoder()

# Iterate over all the values of each column and extract their dtypes
for col in df0:
    # Compare if the dtype is object
    if df0[col].dtype=='object':
    # Use LabelEncoder to do the numeric transformation
        df0[col]=le.fit_transform(df0[col])

In [18]:
df0['applicant_id']

0      1469590
1      1203873
2      1432761
3      1207582
4      1674436
        ...   
995    1880194
996    1114064
997    1758046
998    1824545
999    1660770
Name: applicant_id, Length: 1000, dtype: int64

In [19]:
df0.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 0 to 999
Data columns (total 27 columns):
 #   Column                                                    Non-Null Count  Dtype
---  ------                                                    --------------  -----
 0   loan_application_id                                       1000 non-null   int32
 1   applicant_id                                              1000 non-null   int64
 2   Months_loan_taken_for                                     1000 non-null   int64
 3   Purpose                                                   1000 non-null   int32
 4   Principal_loan_amount                                     1000 non-null   int64
 5   EMI_rate_in_percentage_of_disposable_income               1000 non-null   int64
 6   Property                                                  1000 non-null   int32
 7   Has_coapplicant                                           1000 non-null   int64
 8   Has_guarantor                          

## Upsampling the minor dataset

In [20]:
non_default = df0[df0["high_risk_applicant"] == 0]
default  = df0[df0["high_risk_applicant"] == 1]
print(non_default.shape)
print(default.shape)

(700, 27)
(300, 27)


In [21]:
from sklearn.utils import resample
default = resample(default,
             replace=True,
             n_samples=len(non_default),
             random_state=42)

print(default.shape)
print(type(default))

(700, 27)
<class 'pandas.core.frame.DataFrame'>


### joining the processed data

In [22]:
dfn=pd.concat((non_default, default)).sort_index(axis = 0)

In [23]:
dfn = dfn.drop('applicant_id', axis=1)

In [24]:
dfn.shape

(1400, 26)

In [25]:
dfn['high_risk_applicant'].value_counts()

0    700
1    700
Name: high_risk_applicant, dtype: int64

## Splitting the dataset into train and test sets

In [26]:
from sklearn.model_selection import train_test_split
#convert the DataFrame to a NumPy array
dfn = dfn.values

# Segregate features and labels into separate variables
X,y = dfn[:,0:25] , dfn[:,-1]

# Split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X,
                                y,
                                test_size=0.3,
                                random_state=1)

## Normalising the data

In [27]:
from sklearn.preprocessing import MinMaxScaler

# Instantiate MinMaxScaler and use it to rescale X_train and X_test
scaler = MinMaxScaler(feature_range=(0, 1))
rescaledX_train = scaler.fit_transform(X_train)
rescaledX_test = scaler.fit_transform(X_test)

## Model building ,Making predictions and evaluating performance

In [28]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, f1_score
#from imblearn.over_sampling import RandomOverSampler
#from imblearn.under_sampling import RandomUnderSampler
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, KFold, cross_val_score

In [29]:
log = LogisticRegression(class_weight='balanced')
forest = RandomForestClassifier(n_estimators=150)
svm = SVC()


In [30]:
log_pipe = Pipeline([
    ('scale', StandardScaler()),
    ('log', log)
])

In [31]:
for model, model_name in zip([log_pipe,forest,svm],['Logistic Regression','Random Forest Classifier','SV Classifier']):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print(model_name+":\n")
    print("Accuracy Score:", accuracy_score(y_test, y_pred))
    print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
    print("Clasisfication Report:\n", classification_report(y_test, y_pred))

Logistic Regression:

Accuracy Score: 0.7214285714285714
Confusion Matrix:
 [[152  65]
 [ 52 151]]
Clasisfication Report:
               precision    recall  f1-score   support

           0       0.75      0.70      0.72       217
           1       0.70      0.74      0.72       203

    accuracy                           0.72       420
   macro avg       0.72      0.72      0.72       420
weighted avg       0.72      0.72      0.72       420

Random Forest Classifier:

Accuracy Score: 0.8857142857142857
Confusion Matrix:
 [[183  34]
 [ 14 189]]
Clasisfication Report:
               precision    recall  f1-score   support

           0       0.93      0.84      0.88       217
           1       0.85      0.93      0.89       203

    accuracy                           0.89       420
   macro avg       0.89      0.89      0.89       420
weighted avg       0.89      0.89      0.89       420

SV Classifier:

Accuracy Score: 0.5857142857142857
Confusion Matrix:
 [[173  44]
 [130  73]]
Cl

<b>Observation:</b>
1. Gotten F1-Score of 0.88 for Random forest which is highest.
2. But we have gotten High false negative (actual -- default but predicted -- non-default)  compared to false positive.

## By cross validation process

In [32]:
fold = KFold(n_splits = 5, random_state=1, shuffle = True)
print("Model Evaluation Scores:\n\n")
for model, model_name in zip([log_pipe,forest,svm],['Logistic Regression','Random Forest Classifier','SV Classifier']):
    scores = cross_val_score(model,
                                X_train,
                                y_train,
                                cv = fold,
                                scoring = 'accuracy')
    print(model_name+":")
    print("Accuracy: ",np.mean(scores))
    print("\n")

Model Evaluation Scores:


Logistic Regression:
Accuracy:  0.7020408163265306


Random Forest Classifier:
Accuracy:  0.8704081632653061


SV Classifier:
Accuracy:  0.5469387755102041




In [39]:
for model, model_name in zip([log_pipe,forest,svm],['Logistic Regression','Random Forest Classifier','SV Classifier']):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print(model_name+":\n")
    print("Accuracy Score:", accuracy_score(y_test, y_pred))
    print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
    print("Clasisfication Report:\n", classification_report(y_test, y_pred))

Logistic Regression:

Accuracy Score: 0.7214285714285714
Confusion Matrix:
 [[152  65]
 [ 52 151]]
Clasisfication Report:
               precision    recall  f1-score   support

           0       0.75      0.70      0.72       217
           1       0.70      0.74      0.72       203

    accuracy                           0.72       420
   macro avg       0.72      0.72      0.72       420
weighted avg       0.72      0.72      0.72       420

Random Forest Classifier:

Accuracy Score: 0.8857142857142857
Confusion Matrix:
 [[185  32]
 [ 16 187]]
Clasisfication Report:
               precision    recall  f1-score   support

           0       0.92      0.85      0.89       217
           1       0.85      0.92      0.89       203

    accuracy                           0.89       420
   macro avg       0.89      0.89      0.89       420
weighted avg       0.89      0.89      0.89       420

SV Classifier:

Accuracy Score: 0.5857142857142857
Confusion Matrix:
 [[173  44]
 [130  73]]
Cl

## Implementing RF 

In [37]:
m = RandomForestClassifier(n_estimators=150)
m.fit(X_train, y_train)
y_pred = m.predict(X_test)

In [38]:
print("Accuracy Score:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Clasisfication Report:\n", classification_report(y_test, y_pred))

Accuracy Score: 0.8857142857142857
Confusion Matrix:
 [[183  34]
 [ 14 189]]
Clasisfication Report:
               precision    recall  f1-score   support

           0       0.93      0.84      0.88       217
           1       0.85      0.93      0.89       203

    accuracy                           0.89       420
   macro avg       0.89      0.89      0.89       420
weighted avg       0.89      0.89      0.89       420



<b> Random forest has given us the best result of 0.88 and 0.89 